### DISCUSSION TITLE
next/link vs. next/route.. when to use what?  and how to open a new window after an action?

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#36926] [Help]
next/link and next/route look very similar to me, at glance they seem like different ways to do the same thing which is send people to an href..

So, what are the main differences of next/link vs. next/route? and when would it be better to use what? 

What should I use to link a button to a pdf? 

What should I use on a navbar to go from page to page?

What should I use to go to an external page? 

Oh BTW....  how can I make them both open a news webpage window after a click?

**Comment by icyJoseph** [comment]
Hi,


There's no better/worse between, they are just different utilities of the same API, for very different use cases.

> Check out this 120 lines router, to see why both API's are necessary https://github.com/ashok-khanna/react-snippets/blob/main/Router.js

## Link from next/link

These are a drop in replacement for anchor tags, `<a/>`, to which you'd normally pass a destination URL as `href`.

These exist to let the main router logic know about changes to the history. This is where it gets a bit complicated, but essentially, SPA transitions/routing, need you to use their API, otherwise, the value proposition won't work out. 

When using `ISR`, links will prefetch data for those pages when the `Link` comes into the viewport, but that's perhaps not something you are using yet.

To sum up, a `Link`, is just that, a common link.

## Use router.push, replace and the bunch

The entire SPA experience is exposed to the programmer through this API. Is it possible to implement, somewhat, the `Link` component with this API, yes, but why re-invent the wheel. Also remember that `a` tags convey HTML semantic meaning, and the `Link` component is a simple API to also let the router know about URL changes.

While the `Link` is declarative and built-into the JSXm the methods push and replace, are more of an imperative API. And you can use these to respond to callback success/failure, to prevent access to certain router from the client, etc.

## Learning

To really grasp these, you ought to use them and read the docs, there's no shortcuts.

> What should I use to link a button to a pdf?

A simple `a` tag with the download attribute is good enough.

> What should I use on a navbar to go from page to page?

Use `Link` tags, because these will prevent page unload (SPA transitions), and let the next/router know that the rendered page has to change.

> What should I use to go to an external page?

A `Link` or `a` tag would suffice. Also imperatively, for example after fetching some data, you can use `push`, `replace`. 

> Oh BTW.... how can I make them both open a news webpage window after a click?

Pass `_target="blank"` to the anchor tag. It can be wrapped with a Link or not, doesn't really matter, which is why, doing this with `router` is a noop. 

To do this imperatively, use `window.open`, as in `window.open(url, '_blank').focus();`

Do note, that sometimes, browsers prevent imperative `pop-ups`. Links are considered to be more semantic.

## Gotchas!

I'd prefer to use `Link` whenever links are well-defined, because not everyone has `JavaScript` activated, so using `router.push` or the rest, would make the components using them useless. `Link`, actually does render an anchor tag with an href to the DOM, making it so that people with `JavaScript` disabled, can still navigate and use my page as any other, though, the SPA transition feel would be gone, it is a good trade-off.

The `push`, `replace` methods, are async.
